src/main/ruby/core/truffle/cext.rb

```diff
@@ -1268,6 +1268,22 @@ def rb_hash_set_ifnone(hash, value)
     hash.default = value
   end
 
+  ST_CONTINUE = 0
+  ST_STOP = 1
+  ST_DELETE = 2
+
+  def rb_hash_foreach(hash, func, farg)
+    hash.each do |key, value|
+      st_result = foreign_call_with_block(func, key, value, farg)
+
+      case st_result
+      when ST_CONTINUE then next
+      when ST_STOP then break
+      when ST_DELETE then hash.delete(key)
```

Comment:
I was matching MRI, which doesn't have an `else`. I admit that looks sloppy, but I didn't want to veer too far off course. Having said that, I'll add something.

Reply:
It seems better to check and if it proves problematic we can see then and see the problem very clearly, rather than some obscure bug e.g. due to integers of different width.